### PR TITLE
Create `BaseReconciler`

### LINKED ISSUE #1032: Create `BaseReconciler`
### 🚀 Feature Request

Create class to perform forecast reconcilation

### Proposal

Create class:
```
class BaseReconciler(ABC, BaseMixin):

    def __init__(self, target_level: str, source_level: str):
        """
        target_level: level to be reconciled from the forecasts
        source_level: level to be forecasted
        """
        self.mapping_matrix: Optional[scipy.sparse.base.spmatrix]
```
1. Add abstract method `fit(self, ts: TSDataset) -> "BaseReconciler"` 
- Receive dataset on the the level which is lower or equal than `target_level`, `source_level`
- Aggregate the dataset to the `source_level`
- Fit the parameters of the Reconciler(`mapping_matrix`) -- will differs in children 
2. Add method `aggregate(self, ts: TSDataset) -> "TSDataset"`
- Receive dataset on the the level which is lower or equal than `source_level`
- Aggregate the dataset to the `source_level`
3. Add method `reconcile(self, ts: TSDataset) -> TSDataset:`
- Receive dataset on `source_level`
- Reconcile the forecasts form `ts` using `mapping_matrix` to the `target_level`
  - Y_target =  G * Y_source.T
- Create dataset with new df:
```
TSDataset(df=Y_target, freq=ts.freq, ...)
```

### Test cases

1. Method `aggregate` work correctly for different `source_level`(lower/equal/greater than level of ts) 
2. Method `reconcile` works correctly for top_down/bottom_up mapping matrixes and different level of input ts(lower/equal/greater than `source_level`) 

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by alex-hse-repository**
## Before submitting (must do checklist)

- [ ] Did you read the [contribution guide](https://github.com/tinkoff-ai/etna/blob/master/CONTRIBUTING.md)?
- [ ] Did you update the docs? We use Numpy format for all the methods and classes.
- [ ] Did you write any new necessary tests?
- [ ] Did you update the [CHANGELOG](https://github.com/tinkoff-ai/etna/blob/master/CHANGELOG.md)?

## Closing issues

closes #1032

### CODE REVIEW

File: etna/reconciliation/base.py [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+        if self.mapping_matrix is None:
+            raise ValueError(f"Reconciliator is not fitted!")
+
+        if ts.hierarchical_structure is None:
+            raise ValueError(f"Passed dataset has not hierarchical structure!")
+
+        if ts.current_df_level != self.source_level:
+            raise ValueError(f"Dataset should be on the {self.source_level} level!")
+
+        current_level_segments = ts.hierarchical_structure.get_level_segments(level_name=self.source_level)
```
Comment (brsnw250) [inline etna/reconciliation/base.py:83]: Maybe move this lines before check? In such case, we will check that target and source levels are parts of the hierarchy and, after that, we will check that dataset at the valid level.

File: etna/reconciliation/base.py [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+
+        Returns
+        -------
+        :
+            TSDataset on the ``source_level``.
+        """
+        ts_aggregated = ts.get_level_dataset(target_level=self.source_level)
+        return ts_aggregated
+
+    def reconcile(self, ts: TSDataset) -> TSDataset:
```
Comment (brsnw250) [inline etna/reconciliation/base.py:61]: I think here we also need to check if the target level is higher than that source.

File: etna/reconciliation/base.py [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+            raise ValueError(f"Passed dataset has not hierarchical structure!")
+
+        if ts.current_df_level != self.source_level:
+            raise ValueError(f"Dataset should be on the {self.source_level} level!")
+
+        current_level_segments = ts.hierarchical_structure.get_level_segments(level_name=self.source_level)
+        target_level_segments = ts.hierarchical_structure.get_level_segments(level_name=self.target_level)
+
+        current_level_values = ts[:, current_level_segments, "target"].values
+        target_level_values = current_level_values @ self.mapping_matrix.T
```
Comment (Mr-Geekman) [inline etna/reconciliation/base.py:87]: Will it work properly with series that starts at different timestamps?

File: tests/test_reconciliation/test_base.py [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+def test_aggregate(hierarchical_ts, source_level):
+    reconciliator = DummyReconciliator(target_level="level", source_level=source_level)
+    ts_aggregated = reconciliator.aggregate(ts=hierarchical_ts)
+    assert ts_aggregated.current_df_level == source_level
+
+
+def test_aggregate_fails_low_source_level(hierarchical_ts):
+    ts_market_level = hierarchical_ts.get_level_dataset(target_level="market")
+    reconciliator = DummyReconciliator(target_level="level", source_level="product")
+    with pytest.raises(ValueError, match=""):
```
Comment (Mr-Geekman) [inline tests/test_reconciliation/test_base.py:69]: Can't we check error message more specifically?

File: tests/test_reconciliation/test_base.py [hunk 5]
```diff
@@ -0,0 +1,105 @@
...
+        _ = reconciliator.aggregate(ts=ts_market_level)
+
+
+def test_reconcile_not_fitted_fails(hierarchical_ts):
+    reconciliator = DummyReconciliator(target_level="level", source_level="product")
+    with pytest.raises(ValueError, match=f"Reconciliator is not fitted!"):
```
Comment (Mr-Geekman) [inline tests/test_reconciliation/test_base.py:75]: Do we need f-string here?

**Review by Mr-Geekman — CHANGES_REQUESTED** [review]
Little issues, look at comments above.

File: etna/reconciliation/base.py (hunk 1 above)
Comment (alex-hse-repository) [reply etna/reconciliation/base.py:83]: Call of this methods requires some movements inside)
I think it is better to check the more explicit requirements to the input data(has hierarchy/current level = source level...)
Also the code structure is better now as all the checks are at the beginning and are separate from the other logic

File: etna/reconciliation/base.py (hunk 2 above)
Comment (alex-hse-repository) [reply etna/reconciliation/base.py:61]: It is not required here as this method is common for both top-down and bottom-up

File: etna/reconciliation/base.py (hunk 3 above)
Comment (alex-hse-repository) [reply etna/reconciliation/base.py:87]: It should, we do the same thing [here](https://github.com/tinkoff-ai/etna/blob/7757abe6ca3a7c992b2b2d7d062818b2359c3ca8/etna/datasets/tsdataset.py#L1014) and have tests with nans

Comment (alex-hse-repository) [reply etna/reconciliation/base.py:87]: That is why I thought of moving this logic to get_level_dataset as it is almost the same

File: tests/test_reconciliation/test_base.py (hunk 4 above)
Comment (alex-hse-repository) [reply tests/test_reconciliation/test_base.py:69]: My fall)

File: tests/test_reconciliation/test_base.py (hunk 5 above)
Comment (alex-hse-repository) [reply tests/test_reconciliation/test_base.py:75]: We don't)

**Review by Mr-Geekman — APPROVED** [review]

**Pull request merged**